File: adapters/python/logical_value.py
"""Parses this project's Neutral Logical Value Grammar (schema:
logical-value.schema.json at the repo root) from JSON stdin lines into a
closed set of dataclasses -- mirrors the Go adapter's LogicalValue
interface / Kotlin's sealed LogicalValue class.

Unlike the Go/Kotlin/TypeScript/Rust ports, no special "use an
arbitrary-precision JSON number type" trick is needed here: Python's `json`
module already parses JSON integer literals as native Python `int`, which is
arbitrary-precision, so the "tag" field (up to 2**64-1) is never narrowed
through a lossy fixed-width intermediate."""
import json
from dataclasses import dataclass, field
from typing import List, Union

from errors import ParseError

MAX_UINT64 = (1 << 64) - 1


@dataclass(frozen=True)
class IntValue:
    value: str


@dataclass(frozen=True)
class FloatValue:
    width: str
    value: str


@dataclass(frozen=True)
class TextValue:
    value: str


@dataclass(frozen=True)
class BytesValue:
    value: str


@dataclass(frozen=True)
class BoolValue:
    value: bool


@dataclass(frozen=True)
class NullValue:
    pass


@dataclass(frozen=True)
class ArrayValue:
    items: List["LogicalValue"] = field(default_factory=list)


@dataclass(frozen=True)
class MapEntry:
    key: "LogicalValue"
    val: "LogicalValue"


@dataclass(frozen=True)
class MapValue:
    entries: List[MapEntry] = field(default_factory=list)


@dataclass(frozen=True)
class TagValue:
    tag: int
    value: "LogicalValue"


@dataclass(frozen=True)
class BignumValue:
    sign: str
    value: str


LogicalValue = Union[
    IntValue, FloatValue, TextValue, BytesValue, BoolValue, NullValue,
    ArrayValue, MapValue, TagValue, BignumValue,
]


def parse_logical_value_line(line: str) -> LogicalValue:
    try:
        raw = json.loads(line)
    except json.JSONDecodeError as e:
        raise ParseError(f"malformed JSON: {e}") from e
    return parse_logical_value(raw)


def parse_logical_value(raw) -> LogicalValue:
    if not isinstance(raw, dict):
        raise ParseError("expected a JSON object")
    t = raw.get("type")
    if not isinstance(t, str):
        raise ParseError('missing "type" field')

    if t == "int":
        return IntValue(_required_string(raw, "value", "int"))
    if t == "float":
        return FloatValue(
            _required_string(raw, "width", "float"),
            _required_string(raw, "value", "float"),
        )
    if t == "text":
        return TextValue(_required_string(raw, "value", "text"))
    if t == "bytes":
        return BytesValue(_required_string(raw, "value", "bytes"))
    if t == "bool":
        v = raw.get("value")
        if not isinstance(v, bool):
            raise ParseError('bool: missing "value"')
        return BoolValue(v)
    if t == "null":
        return NullValue()
    if t == "array":
        items = raw.get("items")
        if not isinstance(items, list):
            raise ParseError('array: missing "items"')
        return ArrayValue([parse_logical_value(it) for it in items])
    if t == "map":
        entries = raw.get("entries")
        if not isinstance(entries, list):
            raise ParseError('map: missing "entries"')
        out = []
        for e in entries:
            if not isinstance(e, list) or len(e) != 2:
                raise ParseError("map entry must be a 2-element array")
            out.append(MapEntry(parse_logical_value(e[0]), parse_logical_value(e[1])))
        return MapValue(out)
    if t == "tag":
        raw_tag = raw.get("tag")
        # bool is a subclass of int in Python -- exclude it explicitly so a
        # stray JSON `true`/`false` isn't silently accepted as a tag number.
        if raw_tag is None or isinstance(raw_tag, bool) or not isinstance(raw_tag, int):
            raise ParseError('tag: missing "tag" number')
        if raw_tag < 0 or raw_tag > MAX_UINT64:
            raise ParseError('tag: missing "tag" number')
        if "value" not in raw or raw.get("value") is None:
            raise ParseError('tag: missing "value"')
        inner = parse_logical_value(raw["value"])
        return TagValue(raw_tag, inner)
    if t == "bignum":
        return BignumValue(
            _required_string(raw, "sign", "bignum"),
            _required_string(raw, "value", "bignum"),
        )
    raise ParseError(f"unknown logical-value type: {t!r}")


def _required_string(raw: dict, field_name: str, type_name: str) -> str:
    v = raw.get(field_name)
    if not isinstance(v, str):
        raise ParseError(f'{type_name}: missing "{field_name}"')
    return v

File: adapters/go/logicalvalue.go
package main

import (
	"encoding/json"
	"fmt"
	"math/big"
	"strings"
)

// LogicalValue mirrors the Kotlin adapter's sealed LogicalValue class: a
// closed set of concrete types implementing a marker method, parsed from
// this project's Neutral Logical Value Grammar (logical-value.schema.json).
type LogicalValue interface {
	isLogicalValue()
}

type IntValue struct{ Value string }
type FloatValue struct{ Width, Value string }
type TextValue struct{ Value string }
type BytesValue struct{ Value string }
type BoolValue struct{ Value bool }
type NullValue struct{}
type ArrayValue struct{ Items []LogicalValue }
type MapEntry struct{ Key, Val LogicalValue }
type MapValue struct{ Entries []MapEntry }
type TagValue struct {
	Tag   uint64
	Value LogicalValue
}
type BignumValue struct{ Sign, Value string }

func (IntValue) isLogicalValue()    {}
func (FloatValue) isLogicalValue()  {}
func (TextValue) isLogicalValue()   {}
func (BytesValue) isLogicalValue()  {}
func (BoolValue) isLogicalValue()   {}
func (NullValue) isLogicalValue()   {}
func (ArrayValue) isLogicalValue()  {}
func (MapValue) isLogicalValue()    {}
func (TagValue) isLogicalValue()    {}
func (BignumValue) isLogicalValue() {}

// ParseError is raised for any malformed logical-value JSON line.
type ParseError struct{ msg string }

func (e *ParseError) Error() string { return e.msg }

func parseErrf(format string, args ...any) error {
	return &ParseError{msg: fmt.Sprintf(format, args...)}
}

var maxUint64Big = new(big.Int).SetUint64(^uint64(0))

// parseLogicalValueLine decodes a single JSON stdin line into a LogicalValue.
// json.Number is used throughout so the "tag" field (a JSON number that can
// exceed a signed 64-bit type's range, up to 2^64-1) is never narrowed
// through a lossy intermediate.
func parseLogicalValueLine(line string) (LogicalValue, error) {
	dec := json.NewDecoder(strings.NewReader(line))
	dec.UseNumber()
	var raw any
	if err := dec.Decode(&raw); err != nil {
		return nil, parseErrf("malformed JSON: %v", err)
	}
	return parseLogicalValue(raw)
}

func parseLogicalValue(raw any) (LogicalValue, error) {
	obj, ok := raw.(map[string]any)
	if !ok {
		return nil, parseErrf("expected a JSON object")
	}
	t, ok := obj["type"].(string)
	if !ok {
		return nil, parseErrf("missing \"type\" field")
	}
	switch t {
	case "int":
		v, err := requiredString(obj, "value", "int")
		if err != nil {
			return nil, err
		}
		return IntValue{Value: v}, nil
	case "float":
		w, err := requiredString(obj, "width", "float")
		if err != nil {
			return nil, err
		}
		v, err := requiredString(obj, "value", "float")
		if err != nil {
			return nil, err
		}
		return FloatValue{Width: w, Value: v}, nil
	case "text":
		v, err := requiredString(obj, "value", "text")
		if err != nil {
			return nil, err
		}
		return TextValue{Value: v}, nil
	case "bytes":
		v, err := requiredString(obj, "value", "bytes")
		if err != nil {
			return nil, err
		}
		return BytesValue{Value: v}, nil
	case "bool":
		v, ok := obj["value"].(bool)
		if !ok {
			return nil, parseErrf("bool: missing \"value\"")
		}
		return BoolValue{Value: v}, nil
	case "null":
		return NullValue{}, nil
	case "array":
		items, ok := obj["items"].([]any)
		if !ok {
			return nil, parseErrf("array: missing \"items\"")
		}
		out := make([]LogicalValue, len(items))
		for i, it := range items {
			v, err := parseLogicalValue(it)
			if err != nil {
				return nil, err
			}
			out[i] = v
		}
		return ArrayValue{Items: out}, nil
	case "map":
		entries, ok := obj["entries"].([]any)
		if !ok {
			return nil, parseErrf("map: missing \"entries\"")
		}
		out := make([]MapEntry, len(entries))
		for i, e := range entries {
			pair, ok := e.([]any)
			if !ok || len(pair) != 2 {
				return nil, parseErrf("map entry must be a 2-element array")
			}
			k, err := parseLogicalValue(pair[0])
			if err != nil {
				return nil, err
			}
			v, err := parseLogicalValue(pair[1])
			if err != nil {
				return nil, err
			}
			out[i] = MapEntry{Key: k, Val: v}
		}
		return MapValue{Entries: out}, nil
	case "tag":
		rawTag, ok := obj["tag"]
		if !ok || rawTag == nil {
			return nil, parseErrf("tag: missing \"tag\" number")
		}
		num, ok := rawTag.(json.Number)
		if !ok {
			return nil, parseErrf("tag: missing \"tag\" number")
		}
		big, ok := new(big.Int).SetString(num.String(), 10)
		if !ok || big.Sign() < 0 || big.Cmp(maxUint64Big) > 0 {
			return nil, parseErrf("tag: missing \"tag\" number")
		}
		rawValue, ok := obj["value"]
		if !ok || rawValue == nil {
			return nil, parseErrf("tag: missing \"value\"")
		}
		inner, err := parseLogicalValue(rawValue)
		if err != nil {
			return nil, err
		}
		return TagValue{Tag: big.Uint64(), Value: inner}, nil
	case "bignum":
		sign, err := requiredString(obj, "sign", "bignum")
		if err != nil {
			return nil, err
		}
		v, err := requiredString(obj, "value", "bignum")
		if err != nil {
			return nil, err
		}
		return BignumValue{Sign: sign, Value: v}, nil
	default:
		return nil, parseErrf("unknown logical-value type: %q", t)
	}
}

func requiredString(obj map[string]any, field, typeName string) (string, error) {
	v, ok := obj[field]
	if !ok || v == nil {
		return "", parseErrf("%s: missing %q", typeName, field)
	}
	s, ok := v.(string)
	if !ok {
		return "", parseErrf("%s: missing %q", typeName, field)
	}
	return s, nil
}
